src/com/amazon/ion/impl/IonUTF8.java

```diff
@@ -381,8 +381,9 @@ public final static boolean needsSurrogateEncoding(int unicodeScalar) {
         return (unicodeScalar > MAXIMUM_UTF16_1_CHAR_CODE_POINT);
     }
     public final static char highSurrogate(int unicodeScalar) {
-        assert(unicodeScalar > MAXIMUM_UTF16_1_CHAR_CODE_POINT);
-        assert(unicodeScalar <= Character.MAX_CODE_POINT);
+        if (!(unicodeScalar > MAXIMUM_UTF16_1_CHAR_CODE_POINT) || !(unicodeScalar <= Character.MAX_CODE_POINT)) {
+            throw new IonException("invalid unicodeScalar");
+        }
```

Comment:
I'm guessing that the assertion that was failing was `assert(unicodeScalar <= Character.MAX_CODE_POINT)` for the data in the unsupported encoding, is that correct? Everywhere `highSurrogate` and `lowSurrogate` are called, they are surrounded with a `IonUTF8.needsSurrogateEncoding` check, which verifies that a multi-byte encoding is required. Therefore, the `assert(unicodeScalar > MAXIMUM_UTF16_1_CHAR_CODE_POINT)` can stay as-is because its purpose is to protect ion-java developers from breaking the precondition (which would be considered a bug) without imposing a runtime penalty when the code is running in production (assuming assertions are disabled there).

The `assert(unicodeScalar <= Character.MAX_CODE_POINT)` seems like it is being violated when input data with an unsupported encoding is encountered. This can be converted into an `if` as you've proposed, although I would suggest simplifying it to
```
if (unicodeScalar > Character.MAX_CODE_POINT) {...}
```
and providing a more helpful error message, like
```
"Invalid encoding: encountered non-Unicode character."
```